pkg/kubelet/cm/dra/plugin/plugin.go

```diff
@@ -169,42 +176,37 @@ func (h *RegistrationHandler) validateVersions(
 	if existingPlugin.highestSupportedVersion.LessThan(newPluginHighestVersion) {
 		return newPluginHighestVersion, nil
 	}
-	return nil, errors.New(
-		log(
-			"%s for DRA plugin %q failed. Another plugin with the same name is already registered with a higher supported version: %q",
-			callerName,
-			pluginName,
-			existingPlugin.highestSupportedVersion,
-		),
-	)
-}
-
-func deregisterPlugin(pluginName string) {
-	draPlugins.delete(pluginName)
+	return nil, fmt.Errorf("%s DRA plugin %q: another plugin with the same name is already registered with a higher supported version: %q < %q", what, pluginName, newPluginHighestVersion, existingPlugin.highestSupportedVersion)
 }
 
 // DeRegisterPlugin is called when a plugin has removed its socket,
 // signaling it is no longer available.
 func (h *RegistrationHandler) DeRegisterPlugin(pluginName string) {
-	klog.InfoS("DeRegister DRA plugin", "name", pluginName)
-	deregisterPlugin(pluginName)
-	h.controller.removePlugin(pluginName)
+	if p := draPlugins.delete(pluginName); p != nil {
+		logger := klog.FromContext(p.ctx)
+		logger.V(3).Info("Deregister DRA plugin", "endpoint", p.endpoint)
+
+		// Did exist before. Let's clean up ResourceSlices for it because
+		// it might have died without doing that itself and might not
+		// come back.
+		go h.wipeResourceSlices(pluginName)
```

Comment:
Is there a race when a driver is restarting or upgrading? cc @pohly

For example,

1. the socket file is removed when the driver is dead.
2. the kubelet observes it and starts to swip resource slices in a sperate goroutine.
3. the kubelet fails to clean up the resource slices due to some unexpected errors. At this time, the driver restarted and re-published the resource slices.
4. the kubelet continues to clean up the resource slices, and the new resource slices are removed.

I'm not sure whether the driver will publish the resource slices again after the kubelet removes them in step 4. 

Should we abort the cleaning up process when the driver comes back? I don't know the answer because I'm still learning the dra design.

Reply:
The driver must run a control loop. If the kubelet deletes slices while the driver is running, the driver has to recreate them.

The kubelet is only doing one successful deletecollection, so the two will not "fight" against each other permanently.

Reply:
This answers my previous question related to this.